File: tests/t-prng-seed-native.lua
local th = require"threading"

local function reseed_rounds(n)
  for i = 1, n do
    assert(math.randomseed() == nil)
    local x = math.random()
    local y = math.random(1, 32)
    assert(x >= 0 and x < 1)
    assert(y >= 1 and y <= 32)
  end
  return true
end

assert(reseed_rounds(12))

local workers = {}
for i = 1, 4 do
  workers[i] = th.spawn(function(n)
    return reseed_rounds(n)
  end, 8)
end

for i = 1, #workers do
  local ok, res = workers[i]:join()
  assert(ok == true)
  assert(res == true)
end

File: tests/t-prng-seed-native.c
/*
** no-argument math.randomseed native STOPREQ guard.
*/

#include <assert.h>
#include <stdio.h>

#include "lua.h"
#include "lauxlib.h"
#include "lualib.h"

#include "lib/lua_fixture_helpers.h"
#include "lib/tg_stopreq_fixture_helpers.h"

#include "lj_obj.h"
#include "lj_tg.h"

int main(void)
{
  lua_State *L = ljt_lua_newstate_openlibs();
  TGState *tg;

  tg = G2TG(G(L));
  assert(tg != NULL);

  ljt_tg_set_stopreq(tg);
  ljt_lua_dostring(L,
    "assert(math.randomseed() == nil)\n"
    "local x = math.random()\n"
    "assert(x >= 0 and x < 1)\n");
  ljt_tg_clear_stopreq(tg);
  assert((lj_tg_flags_acq(tg) & TGF_STOPREQ) == 0);

  ljt_lua_dostring(L,
    "assert(math.randomseed() == nil)\n"
    "local y = math.random(1, 32)\n"
    "assert(y >= 1 and y <= 32)\n");

  lua_close(L);
  printf("t-prng-seed-native OK: sticky STOPREQ reseed verified\n");
  return 0;
}
